rabbitsmpp/multiconsumer.go

```diff
@@ -0,0 +1,167 @@
+package rabbitsmpp
+
+import (
+	"errors"
+	"sync"
+
+	"golang.org/x/net/context"
+)
+
+var (
+	errConsumerDuplicate = errors.New("consumer already exist")
+	errConsumerMissing   = errors.New("consumer missing")
+)
+
+// ConsumerContainer is a container that collects the data from all the consumer channels into 1
+// channel it's a fan-in pattern. You can add an remove consumers dynamically
+type ConsumerContainer struct {
+	consumers  map[string]Consumer
+	m          *sync.Mutex
+	wg         *sync.WaitGroup
+	Ctx        context.Context
+	stopper    context.CancelFunc
+	outJobChan chan Job
+	outErrChan chan error
+}
+
+func NewContainer() *ConsumerContainer {
+	consumers := make(map[string]Consumer)
+	var m sync.Mutex
+	var wg sync.WaitGroup
+
+	outJobChan := make(chan Job)
+	outErrChan := make(chan error)
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	return &ConsumerContainer{
+		consumers:  consumers,
+		m:          &m,
+		wg:         &wg,
+		Ctx:        ctx,
+		stopper:    cancel,
+		outJobChan: outJobChan,
+		outErrChan: outErrChan,
+	}
+}
+
+func (container *ConsumerContainer) getNoLock(consumerID string) (Consumer, error) {
+	consumer, ok := container.consumers[consumerID]
+	if !ok {
+		return nil, errConsumerMissing
+	}
+
+	return consumer, nil
+}
+
+func (container *ConsumerContainer) removeNoLock(consumerID string) error {
+
+	if _, ok := container.consumers[consumerID]; !ok {
+		return errConsumerMissing
+	}
+
+	delete(container.consumers, consumerID)
+	return nil
+
+}
+
+func (container *ConsumerContainer) add(consumer Consumer) error {
+	c, _ := container.getNoLock(consumer.ID())
+	if c != nil {
+		return errConsumerDuplicate
+	}
+
+	container.consumers[consumer.ID()] = consumer
+	return nil
+}
+
+func (container *ConsumerContainer) RemoveStop(consumerID string) error {
+	container.m.Lock()
+	defer container.m.Unlock()
+
+	consumer, err := container.getNoLock(consumerID)
+	if err != nil {
+		return err
+	}
+
+	err = consumer.Stop()
+	if err != nil {
+		return err
+	}
+
+	err = container.removeNoLock(consumerID)
+	if err != nil {
+		return nil
+	}
+
+	return nil
+}
+
+func (container *ConsumerContainer) AddRun(consumer Consumer) error {
+	container.m.Lock()
+	defer container.m.Unlock()
+
+	err := container.add(consumer)
+	if err != nil {
+		return err
+	}
+
+	container.wg.Add(1)
+	go container.runConsumer(consumer)
+
+	return nil
+}
+
+func (container *ConsumerContainer) runConsumer(consumer Consumer) {
+	defer container.wg.Done()
```

Comment:
Because, Stop method on the container is waiting for all of the consumers to stop, this way you register all of the consumers via waitgroup.Add  when adding them and when want to wait for them to finish you do wg.Wait()